### DISCUSSION TITLE
MP4 to mp3 conversion issue

### DISCUSSION THREAD

**Discussion by A-MV-J** [aclap-dev/video-downloadhelper#266] [Help and Q&A]
### Your question

Hi All.
Good evening to you. How are you?

I can't convert YouTube videos to MP3 formats according to my requirements. When I create my conversion profile for MP3, the video gets downloaded, but the conversion fails, leaving the video saved on my hard drive. This the error I get

> o is undefined
- moz-extension://5168f3f5-2ea9-48b4-adeb-b992f05605da/background/main.js:42
- aT@moz-extension://5168f3f5-2ea9-48b4-adeb-b992f05605da/background/main.js:42:20605
- _T@moz-extension://5168f3f5-2ea9-48b4-adeb-b992f05605da/background/main.js:46:12485

When using the default/built-in options:

- The "audio only download" only converts to 128 Kbps with 44,000 kHz

- The Mp3 HQ only converts to 256 with 44,000 Khz

I like my MP3 in 320 Kbps with 48,000 kHz. I am forced to use another add-on to get the results I want.

Is there a fix coming? 

### Addon & CoApp version and details

Video DownloadHelper
Version 8.2.1.1 (stable)
Browser locale: en-US
Production build
Built on 2024-02-19
Build options: browser=firefox, linuxlic=false, noyt=false, target=mozilla

Found companion app: VdhCoApp 2.0.19

**Comment by paulrouget** [comment]
Ok, so we're not selecting the highest audio quality stream? Can you please share the youtube url you're using?

**Reply by A-MV-J** [reply]
Hi Paul.
Were you able to find out what caused this issue?

**Comment by A-MV-J** [comment]
This is the YT link: https://www.youtube.com/watch?v=ysy1H77igIA

Below you will find my conversion profile

![Screenshot 2024-03-07 121620](https://github.com/aclap-dev/video-downloadhelper/assets/162514752/2f217e7e-55d6-4a33-90fd-26d30188fa28)

**Reply by ethanwind** [reply]
Hi everybody!
Same issue here on raiplaysound.it domain.

![Screenshot 2024-03-14 195748](https://github.com/aclap-dev/video-downloadhelper/assets/104685114/401afa6b-9fe6-4983-a8ae-35baa3853815)

Video downloadhelper downloads file in mp4 format but it does not convert it automatically.
Also I get troubles to customize MP3HQ output format with 320k quality. I can set it but it won't convert any mp4 to my desired setting with local converter.

Thanks in advance for any help. I will appreciate it.

**Reply by GrampaWildWilly** [reply]
I do not work for VDH.  I'm just another user.

I'm running Windows 11 Pro 64-bit version 23H2 build 22631.3672, Firefox 126.0.1 64-bit, licensed VDH 9.0.2.3a1 beta, CoApp 2.0.19.

For your Justin Timberlake video at https://www.youtube.com/watch?v=ysy1H77igIA, I discovered that there is a 3840x2160 MKV at 60fps variant of this.  I downloaded that & ran ffprobe on it (since the Windows properties dialog is inadequate in what it shows for MKV).  Here's what I got:

[ffprobe.txt](https://github.com/user-attachments/files/15746345/ffprobe.txt)

It turns out the allegedly wonderful video resolution & frame rate are wasted.  The video track here is a still photograph for the duration of the clip.

You do know that if the original input is only, for example, 128Kbps or 192Kbps or 256Kbps, converting it to 320Kbps doesn't actually improve it.  It started as whatever bit rate it was & your conversion to 320Kbps just pads out the file with useless bloat that does not improve the quality of the sound.  Not to mention the fact that bit rates, both audio & video, are generally variable & aren't at the constant level of the numbers reported variously by Windows properties & ffprobe.  How do you determine the bit rate of the original on YouTube?  Where does it show you the audio properties of the clip?

Same idea for the sampling rate.  If the original input is 44,100Hz, you're not gaining anything by converting it to 48,000Hz.  You're just adding useless bloat to the file.

This is not to say there isn't a bug somewhere in the VDH conversion process.  I trust you that you are having trouble with that & @paulrouget Paul needs to look into it.  But are you sure the original was provided at such a high bit rate?  I can attest to VDH successfully downloading content that really is 320Kbps audio.  Try anything reasonably recent at Medici.Tv.  You'll get 320Kbps in & 320Kbps out.  YouTube is well known to compress their content.  So I ask again.  How do you know your content is even being provided at 320Kbps?

**Reply by GrampaWildWilly** [reply]
So I went to raiplay.it & I found mostly content that geoblocked me.  "This content is available only in Italy."  But I kept scrounging around & I stumbled on this:

https://www.raiplay.it/video/2024/04/Belve-Cult-il-meglio-della-stagione---30042024-cf305e91-b378-4823-a353-d7701db47195.html

It looks like some kind of interview show.  Despite the fact that I'm a mad opera lover, I actually do not speak Italian so I have no idea what these people were talking about.  In any case, VDH downloaded that video no problem.  I found an HLS master manifest in the Network Monitor & ran ffprobe on it:

[ffprobe RAI Manifest.txt](https://github.com/user-attachments/files/15747654/ffprobe.RAI.Manifest.txt)

Notice that there is Timed_ID3 data here.  I point this out because it used to cause problems for VDH a while ago.  Today, VDH treats such data correctly: it ignores it completely.  You'll notice in that ffprobe report that ffprobe doesn't know how to interpret it either: Unsupported codec.

The audio track of this video is 44,100Hz.  You might convert this to some lower sampling rate & you will get something not as good.  Any attempt to convert it to a higher sampling rate will give you an item whose sampling rate is 44,100Hz but it will be in a container that could have accommodated, let's say, 48,000Hz if the input had been that high.  But it isn't.  You're never going to get anything better than that.

I tried to use the new VDH 9 UI to do a local file convert.  It just plain flat does not work.  So I hope there's room on your bandwagon.  The function just doesn't do anything when you use the new UI.

So I enabled the old UI.  You must understand, this does not revert VDH to an older release.  It just gives you a UI that accesses VDH 9 using the UI we're used to from past releases.  She's the same modern woman, just wearing her old clothes.  Using the legacy UI, I managed to convert the MP4 I had downloaded into an MP3.  I ran that MP3 through ffprobe:

[ffprobe audio track.txt](https://github.com/user-attachments/files/15747672/ffprobe.audio.track.txt)

Sampling rate still only 44,100Hz, bit rate 128Kbps.  If you try to target anything higher, you will have 44,100Hz/128Kbps content but it will be in a package that could hold something higher.  But it will still be only what it started out as.  You can't increase that.  The only way to get higher numbers is for RAI to have posted something that has higher numbers.  Then you would get something with those higher numbers on your system & you could work with that.

Bottom line: In the modern VDH 9 UI, convert local files is completely non-functional.  If it works, you -- and by that I mean @paulrouget Paul -- have to give us documentation on how to do it.  The software is non-functional until you write natural language documentation for humans to read explaining how to use it.  Documentation is at least as important if not more important than computer software.  This is a concept I will never stop hammering on.  Documentation is entirely more important than you grasp.  I have to believe you tested this before you released it.  I have to believe you got it to do something.  But you haven't told us how to do it.  Until you do, the computer software is non-functional.  It's just a bug taking up space on our systems.  Give us documentation & that might change.  But documentation can't fix software bugs, just as these upconversions that my fellow users here are attempting are a waste of time.  If there are bugs in the software, those also need to be fixed.  I find it strange that enabling the legacy UI allows a conversion to work when the modern UI does nothing.  Both interfaces are accessing the same version of the product.  It's just a weird situation.

**Reply by ethanwind** [reply]
Hi GrampaWildWilly!
First of all I want to thank you for your interest and your response.
I'm using Video DownloadHelper v.9.0.2.3 (stable) on Firefox Developer edition v.127.0b9 (64 bit).
On the raiplaysound.it domain, and not raiplay.it, the automatic conversion from mp4 to mp3 HQ, with the rule preset by me, does not work. At the moment I am using another tool on that domain, in particular I am referring to the podcasts of [Radio 3 RAI](https://www.raiplaysound.it/radio3).
It's a shame because I love Video DownloadHelper. I have been using it for years and years and find it very useful.
Thanks again for your support.

**Reply by GrampaWildWilly** [reply]
I am such a dweeb.  I didn't read your link carefully.  I went to the wrong web site.  Quel stupido.  Is that even Italian?  Maybe it's half French, half Italian.  Fritalian.  Fritalien.  Fritaliano.

Anyway, I decided to try to replicate what you've been talking about.  I made a copy of the MP3 HQ profile that is part of the default configuration supplied by VDH.  Then I adjusted it with a single modification:

![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/deb78ca0-9784-47b9-85c9-19b8f10447fc)

Then I used the legacy UI to do a Download & Convert operation on this item:

https://www.raiplaysound.it/audio/2024/05/Lezioni-di-musica-del-04052024-db481279-3238-41da-8f6b-42939662b790.html

The download succeeded.

![#02](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/9d56d7b9-2d51-4f59-bde0-a5b8996be44c)

Note the audio bit rate there.  This is the file I downloaded.  You can tell that's what it is because it is an MP4.  VDH downloaded this with a bit rate of 320Kbps.  So at least you're starting with the bit rate you were looking for.

I have to say at this point that there doesn't seem to be much of a point in converting this.  It's already got an audio bit rate of 320Kbps.  So what if it's an MP4?  It doesn't have a video track.  But whatever.  You want an MP3, so let's continue.

This is just the download step.  The conversion failed.

![#03](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/d3864985-2b9d-4a86-9220-6abb168f215f)

Like you're surprised.  So I never got to the point of actually doing the conversion.  It failed, like you have all been saying.

However, you don't need to rely on VDH.  VDH actually supplies the tool you need.  In fact, it's the tool that VDH itself uses.  At least, it's the tool it would use if it used it correctly.  The fact that I'm using what VDH itself would use to do the conversion makes it even more baffling that this fails in VDH.  Just do this at a command prompt:

```
"C:\DownloadHelper CoApp\ffmpeg.exe" -fflags +discardcorrupt  -hwaccel auto -protocol_whitelist file,crypto,data,http,https,tls,tcp -strict experimental  -analyzeduration 25000000 -probesize 250000000  -i "M:\VDH Testing\Lezioni di musica S2024E5 Giacomo Puccini, Madama Butterfly - Prima parte Rai Radio 3 RaiPlay Sound.mp4" -ab 320k -y "M:\VDH Testing\Italian Podcast Audio Track.mp3" 2>"M:\VDH Testing\Italian Podcast Audio Track mp3 Log.txt"
```

Some of those parameters are probably unnecessary.  I just based this on the boilerplate I use when I'm applying my Plan B to problems.  You have to scroll that horizontally to see it all.  You can copy/paste the command using the little icon above & to the right of that.

Notice the directory path to the executable.  This is the ffmpeg supplied with VDH, specifically with the CoApp.  I assume VDH would execute a command very much like this, that it will do something like this once the product is fixed.  The key parameter in there is `-ab 320k`.  The default value in ffmpeg is `-ab 128k` which is clearly inadequate for our purposes.  So I had to supply that parameter.  Ffmpeg figures out that you want to convert the MP4 into an MP3 by looking at the extensions in the file names.  The structure of the command is like this:

ffmpeg parameters -i InputFileName parameters OutputFileName WindowsRedirectiontoCapturetheLogFile

For good measure, I ran ffprobe on the result of that ffmpeg run.

```
"C:\DownloadHelper CoApp\ffprobe.exe" -protocol_whitelist file,crypto,data,http,https,tls,tcp -analyzeduration 25000000 -probesize 250000000 -allowed_extensions ALL "M:\VDH Testing\Italian Podcast Audio Track.mp3" 2>>"M:\VDH Testing\Italian Podcast Audio Track mp3 Log.txt"
```

Here's the full results:

[Italian Podcast Audio Track mp3 Log.txt](https://github.com/user-attachments/files/15750777/Italian.Podcast.Audio.Track.mp3.Log.txt)

It isn't very interesting to show VLC playing an audio-only file.  But VLC does provide this information:

![#04](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/eedf3094-665f-4163-aef1-3f865afb63ea)

I'm sure I would listen with rapt attention to this particular podcast . . . if only I spoke Italian.  But I don't, so I just skimmed it to make sure it was all there.

So until our good buddy @paulrouget Paul gets around to fixing this, you can do the conversion without too much difficulty yourself.  There's a few file specifications in there that you would have to adjust to point to directories & files you have on your system.  But otherwise, download (without bothering to convert) whatever you want, then just run those commands.

**Reply by ethanwind** [reply]
Thanks so much, GrampaWildWilly! For your time and your patience and, of course, for your suggestion.
Hopefully, someone at Video DownloadHelper will fix this bug, if reading this conversation.
Thank you again.

**Comment by Nestor-DD** [comment]
Hi, 

Got the same problem converting youtube videos to .mp3.

I'm running
Windows 10 Enterprise LTSC 21H2, OS build 19044.3930
Firefox 123.0.1

Video DownloadHelper
Version 8.2.1.1 (stable)
Browser locale: en-US
Production build
Built on 2024-02-19
Build options: browser=firefox, linuxlic=false, noyt=false, target=mozilla
Licensed VDH MP3 converter

The problem:

- "Audio Only Download": works, creates a 128 kbps .mp3
- "Download & Convert > MP3 HQ": works, creates a 256 kbps .mp3 
- "Download & Convert > MP3 320": fails, creates .mp4 file, gives error:

_o is undefined_

_moz-extension://ef9c3281-57dc-41a2-9c87-64dfa7bae9ab/background/main.js:42_
_aT@moz-extension://ef9c3281-57dc-41a2-9c87-64dfa7bae9ab/background/main.js:42:20605_
__T@moz-extension://ef9c3281-57dc-41a2-9c87-64dfa7bae9ab/background/main.js:46:12485_

I've duplicated the MP3 HQ profile, modified the bitrate to 320 kbps, saved it as "MP3 320 alternate": fails, creates .mp4 file, gives same error.

Addenda:

Do you have a custom arguments reference list for the MP3 converter, so I can set sampling rate, etc...?

**Comment by kd4pba** [comment]
I recall there being an option for this and its no longer available for me. I am stuck with 128kb for all audio
Version 9.5.0.2 (stable), target mozilla, locale en-US.
CachyOS (arch) 64 bit with Companion App installed.
